core/src/main/java/com/linecorp/armeria/client/DefaultDnsQueryLifecycleObserver.java

```diff
@@ -0,0 +1,166 @@
+        UNRECOGNIZED_TYPE,
+        NAME_SERVERS_EXHAUSTED,
+        OTHERS,
+        SERVER_TIMEOUT,
+        RESOLVER_TIMEOUT;
+
+        final String lowerCasedName;
+
+        DnsExceptionTypes() {
+            lowerCasedName = Ascii.toLowerCase(name());
+        }
+    }
+
+    /**
+     * Accepts meterRegistry.
+     * @param meterRegistry {@link MeterRegistry} MeterRegistry to capture metrics.
+     * @param question {@link DnsQuestion} DnsQuestion.
+     */
+    DefaultDnsQueryLifecycleObserver(MeterRegistry meterRegistry,
+                                     DnsQuestion question, MeterIdPrefix prefix) {
+        this.meterRegistry = meterRegistry;
+        meterIdPrefix = prefix;
+        nameTag = Tag.of(NAME_TAG, question.name());
+    }
+
+    @Override
+    public void queryWritten(InetSocketAddress dnsServerAddress, ChannelFuture future) {
+        meterRegistry.counter(meterIdPrefix.name() + ".written",
+                              Arrays.asList(nameTag,
+                              Tag.of(SERVER_TAG, dnsServerAddress.getAddress().getHostAddress())))
+                              .increment();
+    }
+
+    @Override
+    public void queryCancelled(int queriesRemaining) {
+        meterRegistry.counter(meterIdPrefix.name() + ".cancelled",
+                              Tags.of(nameTag)).increment();
+    }
+
+    @Override
+    public DnsQueryLifecycleObserver queryRedirected(List<InetSocketAddress> nameServers) {
+        meterRegistry.counter(meterIdPrefix.name() + ".redirected",
+                              Arrays.asList(nameTag, Tag.of(SERVERS_TAG,
+                              nameServers.stream().map(addr -> addr.getAddress().getHostAddress())
+                              .collect(Collectors.joining(","))))).increment();
+        return this;
+    }
+
+    @Override
+    public DnsQueryLifecycleObserver queryCNAMEd(DnsQuestion cnameQuestion) {
+        meterRegistry.counter(meterIdPrefix.name() + ".cnamed",
+                              Arrays.asList(nameTag,
+                              Tag.of(CNAME_TAG, cnameQuestion.name()))).increment();
+        return this;
+    }
+
+    @Override
+    public DnsQueryLifecycleObserver queryNoAnswer(DnsResponseCode code) {
+        meterRegistry.counter(meterIdPrefix.name() + ".noanswer",
+                              Arrays.asList(nameTag,
+                              Tag.of(CODE_TAG, String.valueOf(code.intValue())))).increment();
+        return this;
+    }
+
+    @Override
+    public void queryFailed(Throwable cause) {
+        meterRegistry.counter(meterIdPrefix.name(),
+                              Arrays.asList(nameTag, Tag.of(RESULT_TAG, "failure"),
+                              Tag.of(CAUSE_TAG, determineDnsExceptionTag(cause).lowerCasedName))).increment();
+    }
+
+    @Override
+    public void querySucceed() {
+        meterRegistry.counter(meterIdPrefix.name(),
+                              Arrays.asList(nameTag, Tag.of(RESULT_TAG, "success"),
```

Comment:
How about creating a singleton instance for `Tag.of(RESULT_TAG, "success")` and reuse?

Reply:
Done